crates/flipperzero/src/filesystem.rs

```diff
@@ -0,0 +1,301 @@
+use core::ffi::{c_char, CStr};
+
+use flipperzero_sys as sys;
+use flipperzero_sys::furi::UnsafeRecord;
+
+const RECORD_STORAGE: *const c_char = sys::c_string!("storage");
+
+/// Stream and file system related error type
+#[derive(Debug, Clone, Copy)]
+pub enum Error {
+    Ok,
+    NotReady,
+    Exists,
+    NotExists,
+    InvalidParameter,
+    Denied,
+    InvalidName,
+    Internal,
+    NotImplemented,
+    AlreadyOpen,
+}
+
+impl Error {
+    pub fn to_sys(&self) -> sys::FS_Error {
+        match self {
+            Self::Ok => sys::FS_Error_FSE_OK,
+            Self::NotReady => sys::FS_Error_FSE_NOT_READY,
+            Self::Exists => sys::FS_Error_FSE_EXIST,
+            Self::NotExists => sys::FS_Error_FSE_NOT_EXIST,
+            Self::InvalidParameter => sys::FS_Error_FSE_INVALID_PARAMETER,
+            Self::Denied => sys::FS_Error_FSE_DENIED,
+            Self::InvalidName => sys::FS_Error_FSE_INVALID_NAME,
+            Self::Internal => sys::FS_Error_FSE_INTERNAL,
+            Self::NotImplemented => sys::FS_Error_FSE_NOT_IMPLEMENTED,
+            Self::AlreadyOpen => sys::FS_Error_FSE_ALREADY_OPEN,
+        }
+    }
+
+    pub fn from_sys(err: sys::FS_Error) -> Self {
+        match err {
+            sys::FS_Error_FSE_OK => Self::Ok,
+            sys::FS_Error_FSE_NOT_READY => Self::NotReady,
+            sys::FS_Error_FSE_EXIST => Self::Exists,
+            sys::FS_Error_FSE_NOT_EXIST => Self::NotExists,
+            sys::FS_Error_FSE_INVALID_PARAMETER => Self::InvalidParameter,
+            sys::FS_Error_FSE_DENIED => Self::Denied,
+            sys::FS_Error_FSE_INVALID_NAME => Self::InvalidName,
+            sys::FS_Error_FSE_INTERNAL => Self::Internal,
+            sys::FS_Error_FSE_NOT_IMPLEMENTED => Self::NotImplemented,
+            sys::FS_Error_FSE_ALREADY_OPEN => Self::AlreadyOpen,
+            _ => unimplemented!(),
+        }
+    }
+}
+
+/// Trait comparable to `std::Read` for the Flipper stream API
+pub trait Read {
```

Comment:
That sounds like a good idea to me. I can't really think of any sensible alternatives.

Should I go ahead and lift the traits, or do you think it's worth trying to get more consensus?

Reply:
That seems reasonable. We can put them under something like `flipperzero::io` similar to the standard library.

If it doesn't work out, it's OK to break the API at the next major release. I'm trying to avoid doing it too often, but given the Flipper Zero SDK API is far from stable, I don't think folks will mind.